added more faqs and answers (#42)
flux-framework/flux-docs: faqs.rst

----------------------

You can read up on reporting bugs here: :ref:`contributing` or report one directly for flux `core <https://github.com/flux-framework/flux-core/issues>`_ or `sched <https://github.com/flux-framework/flux-sched/issues>`_.

.. _not_managing_all_resources:

---------------------------------------------------------------------------------
Why is Flux not discovering and managing all of the resources on the system/node?
---------------------------------------------------------------------------------

* Various bind flags passes to the parallel launcher that started Flux

> * CUDA enabled hwloc (for GPUs)

Reviewer: should include the changes we discussed on the phone:

- give examples for flags used at LLNL for jsrun and srun
- copy the text from the CORAL systems about cuda-enabled hwloc

.. _launch_large_num_jobs:

------------------------------------------------------------
How do I efficiently launch a large number of jobs to Flux?
------------------------------------------------------------

See `bulksumbission <https://github.com/flux-framework/flux-workflow-examples/tree/master/async-bulk-job-submit>`_ for an example workflow or flux `tree <https://github.com/flux-framework/flux-sched/blob/master/src/cmd/flux-tree>`_.

.. _node_memory_exhaustion:

------------------------------------------------------------------
Memory exhaustion on a node when running large ensembles with Flux
------------------------------------------------------------------

> Set ``rundir`` so that the content store is not saved to ``/tmp`` but to filesystem.

Reviewer: I would add an example CLI invocation that sets the rundir.  For example: `flux start -o,-Srundir=/path/to/rundir`.

.. _byop:

-----------------------------
Bring Your Own Python (BYOP)
-----------------------------

Caveats on what will work and what won't swork when using a version of python that isn't the one Flux was configured against.

> 

Reviewer: This should be removed for now until we have a better answer.

.. _mimic_slurm_jobstep:

----------------------------------------
How do I mimic Slurm's jobstep semantics
----------------------------------------

> ``flux mini submit`` ting a script with ``flux mini run`` s will not result in Slurm-styel jobstes unless you stick ``flux-start`` in from the jobscript in ``flux mini sumbit`` .

Reviewer: Spell check

.. _mpi_bootstrap_fails:

---------------------------------------------------------------------------------------------
Flux is failing to bootstrap a specific MPI implementaion (e.g. OpenMPI, MPICH, Spectrum MPI)
---------------------------------------------------------------------------------------------

> Use the ``-o mpi=`` option when running or submitting jobs.

Reviewer: Point to Spectrum MPI subsection in CORAL section.

.. _message_callback_not_run:

-----------------------------------------------------
My message callback is not being run. How do I debug?
-----------------------------------------------------

> * Check the error codes from ``addvec`` , ``register_service`` , the ``RPC`` , etc

Reviewer: I'd add a `TODO` comment that we should link these to the man pages once the flux-core man pages are included in readthedocs.

Reviewer: Unfortunately I was being lazy when jotting these notes down, so they aren't full function names either.  Can we expand them?  For example `RPC` -> `flux_rpc*` and `register_service` -> `flux_register_service`.

> * Use ``FLUX_O_TRACE`` and ``FLUX_HANDLE_TRACE`` to see messages moving through the overlay

Reviewer: I'd make a note that `FLUX_HANDLE_TRACE` is set when starting a flux instance: `FLUX_HANDLE_TRACE=t flux start`

I'd also add a note that `FLUX_O_TRACE` is passed as a flag to `flux_open` with a link to https://github.com/flux-framework/flux-core/blob/9822c63f5e6edf329ab3efb9ce3b8bfe5811e8ab/doc/man3/flux_open.adoc

.. _parallel_run_hang:

------------------------------------------------------------------------------

> I'm experiencig a hang while running my parallel application. How can I debug?

Reviewer: Spell check

------------------------------------------------------------------------------

* Run ``flux mini run/submit`` with the ``-vvv`` argument

> * If it is hanging in startup, try adding the ``PMI DEBUG env var``

Reviewer: ```suggestion
* If it is hanging in startup, try adding the ``PMI_DEBUG`` environment variable: ``PMI_DEBUG=t flux mini run my_app.exe``
```

----------------------

You can read up on reporting bugs here: :ref:`contributing` or report one directly for flux `core <https://github.com/flux-framework/flux-core/issues>`_ or `sched <https://github.com/flux-framework/flux-sched/issues>`_.

.. _not_managing_all_resources:

---------------------------------------------------------------------------------
Why is Flux not discovering and managing all of the resources on the system/node?
---------------------------------------------------------------------------------

> This can be due to various bind flags the need to be passed parallel launcher that started Flux. For example at LLNL you must pass ``--mpibind=off --bind=none`` to ``jsrun``.

Reviewer 2: also: `various bind flags the need to be passed` -> `various bind flags that need to be passed`

Also on all systems, Flux relies on hwloc to auto-detect the on-node resources
available for scheduling.  The hwloc that Flux is linked against must be
configured with ``--enable-cuda`` for Flux to be able to detect Nvidia GPUs.

You can test to see if your system default hwloc is CUDA-enabled with:

.. code-block:: sh

  lstopo | grep CoProc

If no output is produced, then your hwloc is not CUDA-enabled.

.. _launch_large_num_jobs:

------------------------------------------------------------
How do I efficiently launch a large number of jobs to Flux?
------------------------------------------------------------

> See `bulksubmission <https://github.com/flux-framework/flux-workflow-examples/tree/master/async-bulk-job-submit>`_ for an example workflow or flux `tree <https://github.com/flux-framework/flux-sched/blob/master/src/cmd/flux-tree>`_.

Reviewer: ```suggestion
See `bulksubmit.py <https://github.com/flux-framework/flux-workflow-examples/tree/master/async-bulk-job-submit>`_ for an example workflow or `flux tree <https://github.com/flux-framework/flux-sched/blob/master/src/cmd/flux-tree>`_.
```

.. _node_memory_exhaustion:

------------------------------------------------------------------
Memory exhaustion on a node when running large ensembles with Flux
------------------------------------------------------------------

> Set ``rundir`` so that the content store is not saved to ``/tmp`` but to filesystem.

Reviewer: We should add a sentence of context.  Sorry again for the shorthand notes in the FAQ issue:

```suggestion
Flux's in-memory KVS is backed by an on-disk content store.  The default location for the content store is ``/tmp``, which for some systems is configured as a RAMDisk.  To minimize Flux's memory footprint at the cost of a slower content store, set ``rundir`` so that the content store is not saved to ``/tmp`` but to the filesystem.
```

An example command to set the ``rundir`` could look like:

.. code-block:: sh

    flux start -o,-Srundir=/path/to/rundir

.. _mimic_slurm_jobstep:

----------------------------------------
How do I mimic Slurm's jobstep semantics
----------------------------------------

> ``flux mini submit`` ting a script with ``flux mini run`` s will not result in Slurm-style jobsteps unless you stick ``flux-start`` in from the jobscript in ``flux mini submit`` .

Reviewer 3: "in from the jobscript" -> "in front of the jobscript"?

> ``flux mini submit`` ting a script with ``flux mini run`` s will not result in Slurm-style jobsteps unless you stick ``flux-start`` in front of the jobscript in ``flux mini submit`` .

Reviewer: ```suggestion
``flux mini submit`` ting a script with ``flux mini run`` s will not result in Slurm-style jobsteps unless you stick ``flux start`` in front of the jobscript in ``flux mini submit`` .
```

.. _mpi_bootstrap_fails:

----------------------------------------------------------------------------------------------
Flux is failing to bootstrap a specific MPI implementation (e.g. OpenMPI, MPICH, Spectrum MPI)
----------------------------------------------------------------------------------------------

> Use the ``flux mini run/submit -o mpi=`` option when running or submitting.

Reviewer: I would add some additional context here as well.  Something along the lines of "Flux's shell plugins for Intel MPI, MVAPICH, and OpenMPI run by default with every job.  If you experience any issues bootstrapping these MPIs, please open an issue (link to the FAQ on submitting issues).  For Spectrum MPI follow the instructions here: LINK"

> For Spectrum MPI follow the instructions here: :ref:`coral_sched_gpus`

Reviewer: This link points to the GPU section of the CORAL page.  It should point to the subsection above that about running Spectrum MPI.

.. _message_callback_not_run:

-----------------------------------------------------
My message callback is not being run. How do I debug?
-----------------------------------------------------

> * Check the error codes from ``addvec``, ``flux_register_service``, ``flux_rpc*``, etc

Reviewer: ```suggestion
* Check the error codes from ``flux_msg_handler_addvec``, ``flux_register_service``, ``flux_rpc_get``, etc.
```

> * Check the error codes from ``iflux_msg_hadler_addvec``, ``flux_register_service``, ``flux_rpc_get``, etc

Reviewer: ```suggestion
* Check the error codes from ``flux_msg_hadler_addvec``, ``flux_register_service``, ``flux_rpc_get``, etc
```

.. _parallel_run_hang:

-------------------------------------------------------------------------------
I'm experiencing a hang while running my parallel application. How can I debug?
-------------------------------------------------------------------------------

* Run ``flux mini run/submit`` with the ``-vvv`` argument
* If it is hanging in startup, try adding the ``PMI_DEBUG`` environment variable: ``PMI_DEBUG=t flux mini run my_app.exe``

.. _versioning_multi_repo:

-------------------------------------------------------------------
How does the versioning of Flux work with its multi-repo structure?
-------------------------------------------------------------------

For any given repository, the versioning is typical semantic `versioning <https://semver.org/>`_. All of the Flux repos are still < v1.0, so all of our interfaces are subject to change. Once a repo hits v1.0, the interfaces for that repo will only break backwards compatibility on major version increments. New features get added in minor releases. Etc

> The interesting part of the versioning comes from the multi-repo structure. Flux-sched is it's own repo with it's own versioning scheme. A release on flux-core may not break anything in flux-sched or require changes and thus might not warrant a new release. So the flux-core and flux-sched versions do not get incremented in lockstep. Already as of June 2020, flux-core is on 0.16.0 and flux-sched is on 0.8.0. We have the compatibility of the various flux-core/flux-sched versions codified in our `spack packages <https://github.com/spack/spack/blob/5108fe314b92409027c2821698fabb62c0ec3b5d/var/spack/repos/builtin/packages/flux-sched/package.py>`_, and that will get more extensive as we add additional repos like flux-depend and flux-accounting. (TODO: we should make a table and put it in the docs too)

Reviewer: I would comment out this TODO.

```suggestion
The interesting part of the versioning comes from the multi-repo structure. Flux-sched is it's own repo with it's own versioning scheme. A release on flux-core may not break anything in flux-sched or require changes and thus might not warrant a new release. So the flux-core and flux-sched versions do not get incremented in lockstep. Already as of June 2020, flux-core is on 0.16.0 and flux-sched is on 0.8.0. We have the compatibility of the various flux-core/flux-sched versions codified in our `spack packages <https://github.com/spack/spack/blob/5108fe314b92409027c2821698fabb62c0ec3b5d/var/spack/repos/builtin/packages/flux-sched/package.py>`_, and that will get more extensive as we add additional repos like flux-depend and flux-accounting.
.. TODO: we should make a table and put it in the docs too
```
